File: dsa/li01_array/PrefixSum_L560_SubarraySumEqualsK.java
package li01_array;

import java.util.HashMap;

public class PrefixSum_L560_SubarraySumEqualsK {
    public static void main(String[] args) {
        PrefixSum_L560_SubarraySumEqualsK obj = new PrefixSum_L560_SubarraySumEqualsK();
		System.out.println(obj.subarraySum(new int[] { 1,2,3 }, 3));
    }

    // Problem: https://leetcode.com/problems/subarray-sum-equals-k/
    // Idea: Use a map to store prefix sum and its frequency
    // Time: O(n), Space: O(n)
    public int subarraySum(int[] nums, int k) {
        var map = new HashMap<Integer, Integer>();  // sum -> count
        map.put(0, 1);      // if first k elements sum to k, then count = 1

        int sum = 0, cnt = 0;
        for (int num : nums) {
            sum += num;
            int rem = sum - k;
            if (map.containsKey(rem)) {
                cnt += map.get(rem);
            }
            map.putIfAbsent(sum, 0);
            map.put(sum, map.get(sum) + 1);
        }

        return cnt;
    }
}

File: dsa/patterns/01_arrays/python/prefix_sum_l560_subarray_sum_equals_k.py
# PrefixSum L560 - Subarray Sum Equals K
#
# Problem: https://leetcode.com/problems/subarray-sum-equals-k/
# Similar: Count number of Subarray with given xor k
#
# Idea:
# - Use prefix sum with hashmap
# - If (current_sum - k) exists, it contributes to the answer
#
# Time: O(n)
# Space: O(n)


def subarray_sum(nums, k):
    count = 0
    sum = 0
    prefix_count = {0: 1}  # prefix_sum -> frequency

    for num in nums:
        sum += num

        # Check if there exists a prefix sum such that
        # current_sum - prefix_sum = k
        count += prefix_count.get(sum - k, 0)

        # Record current prefix sum
        prefix_count[sum] = prefix_count.get(sum, 0) + 1

    return count


# -------------------- TEST --------------------
if __name__ == "__main__":
    nums = [1, 1, 1]
    k = 2
    print(subarray_sum(nums, k))  # Output: 2
